src/SubHask/Algebra/Matrix.hs

```diff
@@ -0,0 +1,340 @@
+                copyByteArray marr1 off1 arr2 off2 b
+
+        where b = extendDimensions (n2*m2) * Prim.sizeOf (undefined::r)
+
+-- algebra
+instance (Monoid r, Prim r) => Semigroup (Matrix r (m::Symbol) (n::Symbol)) where
+    {-# INLINE (+)  #-} ; (+)  = binopDynU  (+)
+
+instance (Monoid r, Cancellative r, Prim r) => Cancellative (Matrix r (m::Symbol) (n::Symbol)) where
+    {-# INLINE (-)  #-} ; (-)  = binopDynU  (-)
+
+instance (Monoid r, Prim r) => Monoid (Matrix r (m::Symbol) (n::Symbol)) where
+    {-# INLINE zero #-}
+    zero = unsafeInlineIO $ do
+        marr <- V.safeNewByteArray 0 16
+        arr <- unsafeFreezeByteArray marr
+        return $ Matrix_Dynamic arr 0 0 0
+
+instance (Group r, Prim r) => Group (Matrix r (m::Symbol) (n::Symbol)) where
+    {-# INLINE negate #-}
+    negate v = monopDynU negate v
+
+instance (Monoid r, Abelian r, Prim r) => Abelian (Matrix r (m::Symbol) (n::Symbol))
+
+instance (Module r, Prim r) => Module (Matrix r (m::Symbol) (n::Symbol)) where
+    {-# INLINE (.*)   #-} ;  (.*)  v r = monopDynU  (.*r) v
+
+type instance Actor (Matrix r (m::Symbol) (n::Symbol)) = Actor r
+
+instance (Action r, Semigroup r, Prim r) => Action (Matrix r (m::Symbol) (n::Symbol)) where
+  {-# INLINE (.+)   #-}
+  (.+) v r = monopDynU (.+r) v
+
+instance (FreeModule r, Prim r) => FreeModule (Matrix r (m::Symbol) (n::Symbol)) where
+  {-# INLINE (.*.)  #-} ;  (.*.)     = binopDynU  (.*.)
+  ones = undefined
+
+instance (VectorSpace r, Prim r) => VectorSpace (Matrix r (m::Symbol) (n::Symbol)) where
+    {-# INLINE (./)   #-} ;  (./)  v r = monopDynU  (./r) v
+    {-# INLINE (./.)  #-} ;  (./.)     = binopDynU  (./.)
+
+{-# INLINE (!!) #-}
+(!!) :: Prim r => Matrix r (m::Symbol) (n::Symbol) -> (Int, Int) -> r
+(!!) (Matrix_Dynamic arr off _ n) (i,j) = indexByteArray arr (off+i*n+j)
+
+{-# INLINE row #-}
+row :: Prim r => Matrix r (m::Symbol) (n::Symbol) -> Int -> V.UVector (m::Symbol) r
+row (Matrix_Dynamic arr off m _) i = V.UVector_Dynamic arr (off+i*m) m
+
+{-# INLINE col #-}
+col :: ( MatrixField r
+       , V.ValidUVector n r
+       ) => Matrix r (m::Symbol) (n::Symbol) -> Int -> V.UVector (n::Symbol) r
+col m1@(Matrix_Dynamic _ _ m _) j = unsafeToModule [m1!!(i,j) | i <- [0..m-1]]
+
+{-# INLINE mmult #-}
+mmult :: ( MatrixField r
+         , V.ValidUVector x0 r
+         )
+         => Matrix r (m::Symbol) (x0::Symbol)
+         -> Matrix r (x0::Symbol) (n::Symbol)
+         -> Matrix r (m::Symbol) (n::Symbol)
+mmult m1@(Matrix_Dynamic _ _ r1 _) m2@(Matrix_Dynamic _ _ _ c2) =
+  unsafeToModuleM r1 [ m2 `row` i <> m1 `col` j | i <- [0..r1-1], j <- [0..c2-1] ]
+
+transposeM :: ( MatrixField r
+              )
+              => Matrix (Scalar r) (m::Symbol) (n::Symbol)
+              -> Matrix r (n::Symbol) (m::Symbol)
+transposeM m@(Matrix_Dynamic _ _ r c) =
+  unsafeToModuleM c [m!!(j,i) | i <- [0..r-1], j <- [0..c-1] ]
+
+data Direction = Forward | Backward
+
+substitution ::
```

Comment:
I'm not sure what this function is trying to do exactly?  AFAICT, it's not actually used anywhere is it?